kube/src/config/auth.rs

```diff
@@ -0,0 +1,282 @@
+use std::{
+    convert::{TryFrom, TryInto},
+    process::Command,
+    sync::Arc,
+};
+
+use chrono::{DateTime, Utc};
+use http::header;
+use jsonpath_lib::select as jsonpath_select;
+use serde::{Deserialize, Serialize};
+use tokio::sync::Mutex;
+
+use super::{utils, AuthInfo, AuthProviderConfig, ExecConfig};
+use crate::{error::ConfigError, oauth2, Result};
+
+#[derive(Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum Authentication {
+    None,
+    Basic(String),
+    Token(String),
+    RefreshableToken(Arc<Mutex<(String, DateTime<Utc>)>>, AuthInfo),
+}
+
+impl Authentication {
+    pub(crate) async fn to_header(&self) -> Result<Option<header::HeaderValue>, ConfigError> {
+        match self {
+            Self::None => Ok(None),
+            Self::Basic(value) => Ok(Some(
+                header::HeaderValue::from_str(value).map_err(ConfigError::InvalidBasicAuth)?,
+            )),
+            Self::Token(value) => Ok(Some(
+                header::HeaderValue::from_str(value).map_err(ConfigError::InvalidBearerToken)?,
+            )),
+            Self::RefreshableToken(data, auth_info) => {
+                let mut locked_data = data.lock().await;
+                // Add some wiggle room onto the current timestamp so we don't get any race
+                // conditions where the token expires while we are refreshing
+                if chrono::Utc::now() + chrono::Duration::seconds(60) >= locked_data.1 {
+                    if let Authentication::RefreshableToken(d, _) = auth_info.try_into()? {
+                        let (new_token, new_expire) = Arc::try_unwrap(d)
+                            .expect("Unable to unwrap Arc, this is likely a programming error")
+                            .into_inner();
+                        locked_data.0 = new_token;
+                        locked_data.1 = new_expire;
+                    } else {
+                        return Err(ConfigError::UnrefreshableTokenResponse);
+                    }
+                }
+                Ok(Some(
+                    header::HeaderValue::from_str(&locked_data.0).map_err(ConfigError::InvalidBearerToken)?,
+                ))
+            }
+        }
+    }
+}
+
+impl TryFrom<&AuthInfo> for Authentication {
+    type Error = ConfigError;
+
+    /// Loads the authentication header from the credentials available in the kubeconfig. This supports
+    /// exec plugins as well as specified in
+    /// https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins
+    fn try_from(auth_info: &AuthInfo) -> Result<Self, Self::Error> {
```

Comment:
This should be changed to check `auth_info.provider`. That way, we don't need a separate function to get the token from the provider and refreshing will work the same way.